pkg/controller/installation/installation_controller.go

```diff
@@ -386,51 +315,157 @@ func (r *ReconcileInstallation) Reconcile(request reconcile.Request) (reconcile.
+		logrus.Info("updating object")
+		err := r.client.Update(context.TODO(), installation)
+		if err != nil {
+			return err
 		}
-		return reconcile.Result{}, err
 	}
+	return nil
+}
 
-	// UPDATE OBJECT
-	err = r.client.Update(r.context, installation)
+func (r *ReconcileInstallation) handleUninstall(installation *integreatlyv1alpha1.RHMI) (reconcile.Result, error) {
+	retryRequeue := reconcile.Result{
+		Requeue:      true,
+		RequeueAfter: 10 * time.Second,
+	}
+	installationCfgMap := os.Getenv("INSTALLATION_CONFIG_MAP")
+	if installationCfgMap == "" {
+		installationCfgMap = installation.Spec.NamespacePrefix + DefaultInstallationConfigMapName
+	}
+	configManager, err := config.NewManager(context.TODO(), r.client, installation.Namespace, installationCfgMap, installation)
 	if err != nil {
-		// The 'Update' function can error if the resource has been updated by another process and the versions are not correct.
-		if k8serr.IsConflict(err) {
-			// If there is a conflict, requeue the resource and retry Update
-			logrus.Info("Error updating Installation resource. Requeue and retry.", err)
-			return retryRequeue, nil
-		}
+		return reconcile.Result{}, err
+	}
+
+	// Set metrics status to unavailable
+	metrics.RHMIStatusAvailable.Set(0)
+
+	installation.Status.Stage = integreatlyv1alpha1.StageName("deletion")
+	installation.Status.LastError = ""
 
-		logrus.Error(err, "error reconciling installation installation")
-		if installInProgress {
-			return retryRequeue, err
+	// updates rhmi status metric to deletion
+	metrics.SetRHMIStatus(installation)
+
+	// Clean up the products which have finalizers associated to them
+	merr := &multiErr{}
+	finalizers := []string{}
+	for _, finalizer := range installation.Finalizers {
+		finalizers = append(finalizers, finalizer)
+	}
+	for _, productFinalizer := range finalizers {
+		if !strings.Contains(productFinalizer, "integreatly") {
+			continue
 		}
-		return reconcile.Result{}, err
+		productName := strings.Split(productFinalizer, ".")[1]
+		product := installation.GetProductStatusObject(integreatlyv1alpha1.ProductName(productName))
+		reconciler, err := products.NewReconciler(product.Name, r.restConfig, configManager, installation, r.mgr)
+		if err != nil {
+			merr.Add(fmt.Errorf("Failed to build reconciler for product %s: %w", product.Name, err))
+		}
+		serverClient, err := k8sclient.New(r.restConfig, k8sclient.Options{})
+		if err != nil {
+			merr.Add(fmt.Errorf("Failed to create server client for %s: %w", product.Name, err))
+		}
+		phase, err := reconciler.Reconcile(context.TODO(), installation, product, serverClient)
+		if err != nil {
+			merr.Add(fmt.Errorf("Failed to reconcile product %s: %w", product.Name, err))
+		}
+		logrus.Infof("current phase for %s is: %s", product.Name, phase)
 	}
-	if installInProgress {
+
+	//any products left to uninstall, or any errors during product uninstall, add to CR and return
+	if len(installation.Finalizers) > 1 || len(merr.errors) > 0 {
```

Comment:
why `> 1` here? Is 1 finalizer reserved for the intly operator itself or should it be > 0?

Reply:
yes, there is one reserved finalizer for the operator itself.